File: Raven.Client.Java/src/main/java/net/ravendb/client/linq/LinqPathProvider.java
package net.ravendb.client.linq;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.client.document.DocumentConvention;

import org.apache.commons.lang.StringUtils;
import org.codehaus.jackson.annotate.JsonProperty;


import com.mysema.query.types.Constant;
import com.mysema.query.types.Expression;
import com.mysema.query.types.Operation;
import com.mysema.query.types.Ops;
import com.mysema.query.types.Path;
import com.mysema.query.types.PathType;
import com.mysema.query.types.expr.NumberOperation;
import com.mysema.query.types.path.MapPath;

public class LinqPathProvider {
  public static class Result {
    private Class<?> memberType;
    private String path;
    private boolean isNestedPath;
    private Field maybeProperty;

    public Field getMaybeProperty() {
      return maybeProperty;
    }
    public Class< ? > getMemberType() {
      return memberType;
    }
    public String getPath() {
      return path;
    }
    public boolean isNestedPath() {
      return isNestedPath;
    }
    public void setMaybeProperty(Field maybeProperty) {
      this.maybeProperty = maybeProperty;
    }
    public void setMemberType(Class< ? > memberType) {
      this.memberType = memberType;
    }
    public void setNestedPath(boolean isNestedPath) {
      this.isNestedPath = isNestedPath;
    }
    public void setPath(String path) {
      this.path = path;
    }

  }

  private final DocumentConvention conventions;

  public LinqPathProvider(DocumentConvention documentConvention) {
    this.conventions = documentConvention;
  }

  public Result getPath(Expression< ? > expression) {
    expression = simplifyExpression(expression);

    if (expression instanceof NumberOperation<?>) {
      Result customMethodResult = conventions.translateCustomQueryExpression(this, expression);
      if (customMethodResult != null) {
        return customMethodResult;
      }

      NumberOperation<?> numberOperation = (NumberOperation<?>) expression;
      if (numberOperation.getOperator().equals(Ops.MAP_SIZE) || numberOperation.getOperator().equals(Ops.ARRAY_SIZE)) {
        if (numberOperation.getArgs().size() != 1) {
          throw new IllegalArgumentException("Invalid computation: " + numberOperation
            + ". You cannot use computation (only simple member expression are allowed) in RavenDB queries.");
        }
        Result target = getPath(numberOperation.getArg(0));

        Result newResult = new Result();
        newResult.setMemberType(numberOperation.getType());
        newResult.setNestedPath(false);
        newResult.setPath(target.getPath() + ".Count\\(\\)");
        return newResult;
      }

    }

    if (expression instanceof Operation) {
      Operation<?> operation = (Operation< ? >) expression;
      if (operation.getOperator().equals(Ops.STRING_LENGTH) || operation.getOperator().equals(Ops.ARRAY_SIZE)) {
        Result result = getPath(operation.getArg(0));
        result.setPath(result.getPath() + ".Length");
        return result;
      } else if (operation.getOperator().equals(Ops.COL_SIZE)) {
        Result result = getPath(operation.getArg(0));
        result.setPath(result.getPath() + ".Count");
        return result;
      }
    }

    if (expression instanceof Path) {
      Path<?> path = (Path<?>) expression;

      PathType pathType = path.getMetadata().getPathType();
      if (PathType.MAPVALUE_CONSTANT.equals(pathType)) {
        MapPath< ? , ?, ?> mapPath = (MapPath< ? , ? , ? >) path.getMetadata().getParent();
        Result parent = getPath(mapPath);
        Result newResult = new Result();
        newResult.setMemberType(expression.getType());
        newResult.setNestedPath(false);
        newResult.setPath(parent.getPath() + "." + getValueFromExpression(path, mapPath.getKeyType()));

        return newResult;
      }
      if (PathType.PROPERTY.equals(pathType) || PathType.VARIABLE.equals(pathType)) {


        Path< ? > memberExpression = getMemberExpression(expression);
        Result customMemberResult = conventions.translateCustomQueryExpression(this, memberExpression);
        if (customMemberResult != null) {
          return customMemberResult;
        }

        assertNoComputation(memberExpression);

        Result newResult = new Result();

        newResult.setPath(extractPath(memberExpression));
        newResult.setNestedPath(memberExpression.getMetadata().getParent().getMetadata().getPathType().equals(PathType.PROPERTY));
        newResult.setMemberType(memberExpression.getType());


        newResult.setMaybeProperty(fetchMaybeProperty(memberExpression));
        newResult.setPath(handlePropertyRenames(memberExpression, newResult.getPath()));
        return newResult;
      }
    }
    throw new IllegalArgumentException("Don't know how to translate: " + expression);
  }

  private Field fetchMaybeProperty(Path<?> memberExpression) {
    try {
      Path<?> parent = memberExpression.getMetadata().getParent();
      if (parent == null) {
        return null;
      }
      String fieldName = memberExpression.getMetadata().getName();
      Field declaredField = null;
      Class<?> type = parent.getType();
      while (true) {
        declaredField = type.getDeclaredField(fieldName);
        if (declaredField != null) {
          return declaredField;
        }
        type = type.getSuperclass();
        if (Object.class.equals(type)) {
          return null;
        }
      }
    } catch (NoSuchFieldException e) {
      throw new RuntimeException(e);
    }
  }

  private String extractPath(Path<?> expression) {
    List<String> items = new ArrayList<>();

    while (expression != null) {
      if (expression.getMetadata().getPathType() == PathType.DELEGATE) {
        expression = expression.getMetadata().getParent();
        continue;
      }
      items.add(0, StringUtils.capitalize(expression.getMetadata().getName()));
      expression = expression.getMetadata().getParent();
    }
    return StringUtils.join(items, ".");
  }

  public static String handlePropertyRenames(Path<?> member, String name) {
    JsonProperty jsonProperty = member.getAnnotatedElement().getAnnotation(JsonProperty.class);
    if (jsonProperty != null) {
      String propertyName = jsonProperty.value();
      if (StringUtils.isNotEmpty(propertyName)) {
        return name.substring(0, name.length() - member.getMetadata().getName().length()) + propertyName;
      }
    }
    return name;
  }

  private static Expression<?> simplifyExpression(Expression<?> expression) {
    // do nothing
    return expression;
  }

  public Object getValueFromExpression(Expression<?> expression, Class< ? > type) {
    if (expression == null) {
      return new IllegalArgumentException("Value is missing");
    }
    // get object
    Reference<Object> valueRef= new Reference<>();
    if (getValueFromExpressionWithoutConversion(expression, valueRef)) {
      if (valueRef.value == null) {
        return null;
      }
      if (valueRef.value instanceof Enum) {
        if (!conventions.isSaveEnumsAsIntegers()) {
          return ((Enum<?>)valueRef.value).name();
        }
        return ((Enum<?>)valueRef.value).ordinal();
      }

      return valueRef.value;
    }
    throw new IllegalStateException("Can't extract value from expression of type:" + expression);
  }

  public Path<?> getMemberExpression(Expression<?> expression) {
    if (expression instanceof Path) {
      return (Path< ? >) expression;
    }
    throw new IllegalStateException("Could not understand how to translate '" + expression + "' to a RavenDB query.\n" +
      "Are you trying to do computation during the query?\n" +
      "RavenDB doesn't allow computation during the query, computation is only allowed during index. Consider moving the operation to an index."
      );
  }

  public static boolean getValueFromExpressionWithoutConversion(Expression<?> expression, Reference<Object> valueRef) {
    if (expression instanceof Constant) {
      valueRef.value = ((Constant<?>) expression).getConstant();
      return true;
    }
    return false;
  }

  private void assertNoComputation(Path<?> expression) {
    //empty
  }

}

File: src/Raven.Client/Documents/Linq/LinqPathProvider.cs
// -----------------------------------------------------------------------
//  <copyright file="LinqPathProvider.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Raven.Client.Documents.Conventions;
using Raven.Client.Documents.Queries;
using Raven.Client.Documents.Queries.TimeSeries;
using Raven.Client.Documents.Session;
using Raven.Client.Extensions;
using Raven.Client.Util;

namespace Raven.Client.Documents.Linq
{
    public class LinqPathProvider
    {
        public class Result
        {
            public Type MemberType;
            public string Path;
            public bool IsNestedPath;
            public PropertyInfo MaybeProperty;
            public string[] Args;
        }

        private readonly DocumentConventions _conventions;

        public LinqPathProvider(DocumentConventions conventions)
        {
            _conventions = conventions;
        }

        /// <summary>
        /// Get the path from the expression, including considering dictionary calls
        /// </summary>
        public Result GetPath(Expression expression)
        {
            expression = SimplifyExpression(expression);

            if (expression is MethodCallExpression callExpression)
            {
                var customMethodResult = _conventions.TranslateCustomQueryExpression(this, callExpression);
                if (customMethodResult != null)
                    return customMethodResult;

                if (callExpression.Method.Name == "Count" && callExpression.Method.DeclaringType == typeof(Enumerable))
                {
                    if (callExpression.Arguments.Count != 1)
                        throw new ArgumentException("Not supported computation: " + callExpression +
                                            ". You cannot use computation in RavenDB queries (only simple member expressions are allowed).");

                    var target = GetPath(callExpression.Arguments[0]);
                    return new Result
                    {
                        MemberType = callExpression.Method.ReturnType,
                        IsNestedPath = false,
                        Path = target.Path + @".Count"
                    };
                }

                if (callExpression.Method.Name == "get_Item")
                {
                    var parent = GetPath(callExpression.Object);

                    var itemKey = GetValueFromExpression(callExpression.Arguments[0], callExpression.Method.GetParameters()[0].ParameterType).ToString();

                    return new Result
                    {
                        MemberType = callExpression.Method.ReturnType,
                        IsNestedPath = true,
                        Path = parent.Path + "." + itemKey
                    };
                }

                if (callExpression.Method.Name == "ToString" && callExpression.Method.ReturnType == typeof(string))
                {
                    var target = GetPath(callExpression.Object);
                    return new Result
                    {
                        MemberType = typeof(string),
                        IsNestedPath = false,
                        Path = target.Path
                    };
                }

                if (callExpression.Method.Name == "CompareTo" && callExpression.Method.DeclaringType == typeof(string))
                {
                    var target = GetPath(callExpression.Object);
                    return new Result
                    {
                        MemberType = typeof(string),
                        IsNestedPath = false,
                        Path = target.Path
                    };
                }

                if (IsCounterCall(callExpression))
                {
                    return CreateCounterResult(callExpression);
                }

                throw new NotSupportedException(
                    $"Cannot understand how to translate method '{callExpression.Method.Name}' of '{callExpression.Method.DeclaringType}' type. {callExpression}");
            }

            var memberExpression = GetMemberExpression(expression);

            var customMemberResult = _conventions.TranslateCustomQueryExpression(this, memberExpression);
            if (customMemberResult != null)
                return customMemberResult;

            string path;

            switch (memberExpression.Member.Name)
            {
                case "Value":
                    // we truncate the nullable .Value because in json all values are nullable
                    if(Nullable.GetUnderlyingType(memberExpression.Expression.Type) != null)
                        return GetPath(memberExpression.Expression);
                    break;
                case "Values":
                    // if we have a .Keys / .Values we check if we need to omit the dictionary

                    if (memberExpression.Member.DeclaringType != null &&
                        memberExpression.Member.DeclaringType.IsGenericType &&
                        (memberExpression.Member.DeclaringType.GetGenericTypeDefinition() == typeof(Dictionary<,>) ||
                         memberExpression.Member.DeclaringType.GetGenericTypeDefinition() == typeof(IDictionary<,>)))
                    {
                        return GetPath(memberExpression.Expression);
                    }
                    
                    break;
            }


            if (memberExpression.Expression is MemberExpression mi && 
                mi.Member.Name == "Value" &&
                mi.Member.DeclaringType != null && 
                mi.Member.DeclaringType.IsGenericType &&
                mi.Member.DeclaringType.GetGenericTypeDefinition() == typeof(KeyValuePair<,>)
                )
            {
                path = mi.Expression + "." + memberExpression.Member.Name;
            }
            else
            {
                path = memberExpression.ToString();
            }

            AssertNoComputation(memberExpression);

            var result = new Result
            {
                Path = path,
                IsNestedPath = memberExpression.Expression is MemberExpression,
                MemberType = memberExpression.Member.GetMemberType(),
                MaybeProperty = memberExpression.Member as PropertyInfo
            };

            result.Path = HandleMemberExpressionPropertyRenames(memberExpression, result.Path);

            return result;
        }

        internal static Result CreateCounterResult(MethodCallExpression callExpression)
        {
            var counterName = (callExpression.Arguments[callExpression.Arguments.Count - 1] as ConstantExpression)?.Value.ToString();

            string[] args;
            if (callExpression.Method.DeclaringType != typeof(RavenQuery))
            {
                // session.CountersFor().Get()
                var path = (callExpression.Object as MethodCallExpression)?.Arguments[0].ToString();
                args = new[] { RemoveTransparentIdentifiersIfNeeded(path), counterName };
            }
            else if (callExpression.Arguments.Count == 2)
            {
                var path = callExpression.Arguments[0].ToString();
                args = new[] { RemoveTransparentIdentifiersIfNeeded(path), counterName };
            }
            else
            {
                args = new[] { counterName };
            }

            return new Result
            {
                MemberType = typeof(long?),
                IsNestedPath = false,
                Path = "counter",
                Args = args
            };
        }

        private static string HandleMemberExpressionPropertyRenames(MemberExpression memberExpression, string name)
        {
            var member = memberExpression.Member;

            if (memberExpression.Expression is MemberExpression innerMemberExpression)
            {
                var innerName = name.Substring(0, name.Length - member.Name.Length - 1);
                name = HandleMemberExpressionPropertyRenames(innerMemberExpression, innerName);

                if (member.Name != "Value" ||
                    Nullable.GetUnderlyingType(memberExpression.Expression.Type) == null)
                {
                    name += $".{member.Name}";
                }
            }

            return HandlePropertyRenames(member, name);
        }

        public static string HandlePropertyRenames(MemberInfo member, string name)
        {
            var jsonPropAttributes = member.GetCustomAttributes(false)
                                                     .OfType<JsonPropertyAttribute>()
                                                     .ToArray();

            if (jsonPropAttributes.Length != 0)
            {
                string propertyName = jsonPropAttributes[0].PropertyName;
                if (string.IsNullOrEmpty(propertyName) == false)
                {
                    return name.Substring(0, name.Length - member.Name.Length) + propertyName;
                }
            }

            var dataMemberAttributes = member.GetCustomAttributes(false)
                                                       .OfType<DataMemberAttribute>()
                                                       .ToArray();

            if (dataMemberAttributes.Length != 0)
            {
                string propertyName = ((dynamic)dataMemberAttributes[0]).Name;
                if (string.IsNullOrEmpty(propertyName) == false)
                {
                    return name.Substring(0, name.Length - member.Name.Length) + propertyName;
                }
            }
            return name;
        }

        private static Expression SimplifyExpression(Expression expression)
        {
            while (true)
            {
                switch (expression.NodeType)
                {
                    case ExpressionType.Convert:
                    case ExpressionType.ConvertChecked:
                    case ExpressionType.Quote:
                        expression = ((UnaryExpression)expression).Operand;
                        break;
                    case ExpressionType.Lambda:
                        expression = ((LambdaExpression)expression).Body;
                        break;
                    default:
                        return expression;
                }
            }
        }

        /// <summary>
        /// Get the actual value from the expression
        /// </summary>
        public object GetValueFromExpression(Expression expression, Type type)
        {
            if (expression == null)
                throw new ArgumentNullException(nameof(expression));

            // Get object
            if (GetValueFromExpressionWithoutConversion(expression, out var value))
            {
                var nonNullableType = Nullable.GetUnderlyingType(type) ?? type;
                if (value is Enum || nonNullableType.IsEnum)
                {
                    if (value is IEnumerable valueAsEnumerable)
                    {
                        return valueAsEnumerable
                            .Cast<object>()
                            .Select(x => ConvertEnum(x, nonNullableType));
                    }

                    return ConvertEnum(value, nonNullableType);
                }

                return value;
            }

            throw new InvalidOperationException("Can't extract value from expression of type: " + expression.NodeType);

            object ConvertEnum(object val, Type enumType)
            {
                if (val == null)
                    return null;
                if (_conventions.SaveEnumsAsIntegers == false)
                    return Enum.GetName(enumType, val);
                return Convert.ToInt32(val);
            }
        }

        /// <summary>
        /// Get the member expression from the expression
        /// </summary>
        public static MemberExpression GetMemberExpression(Expression expression)
        {
            if (expression is UnaryExpression unaryExpression)
                return GetMemberExpression(unaryExpression.Operand);

            if (expression is LambdaExpression lambdaExpression)
                return GetMemberExpression(lambdaExpression.Body);

            if (!(expression is MemberExpression memberExpression))
            {
                throw new InvalidOperationException("Could not understand how to translate '" + expression + "' to a RavenDB query." +
                                                    Environment.NewLine +
                                                    "Are you trying to do computation during the query?" + Environment.NewLine +
                                                    "RavenDB doesn't allow computation during the query, computation is only allowed during index. Consider moving the operation to an index.");
            }

            return memberExpression;
        }

        public static bool GetValueFromExpressionWithoutConversion(Expression expression, out object value)
        {
            switch (expression.NodeType)
            {
                case ExpressionType.Constant:
                    value = ((ConstantExpression)expression).Value;
                    return true;
                case ExpressionType.MemberAccess:
                    value = GetMemberValue(((MemberExpression)expression));
                    return true;
                case ExpressionType.MemberInit:
                    var memberInitExpression = ((MemberInitExpression)expression);
                    value = Expression.Lambda(memberInitExpression).Compile().DynamicInvoke();
                    return true;
                case ExpressionType.New:
                    value = GetNewExpressionValue(expression);
                    return true;
                case ExpressionType.Lambda:
                    var lambda = ((LambdaExpression)expression);
                    value = lambda.Compile().DynamicInvoke();
                    return true;
                case ExpressionType.Call:
                    if (expression is MethodCallExpression mce)
                    {
                        if (mce.Method.DeclaringType == typeof(RavenQuery) &&
                            mce.Method.Name == nameof(RavenQuery.CmpXchg))
                        {
                            if (TryGetMethodArguments(mce, out var args) == false)
                            {
                                value = null;
                                return false;
                            }
                            value = CmpXchg.Value((string)args[0]);
                            return true;
                        }
                    }
                    value = Expression.Lambda(expression).Compile().DynamicInvoke();
                    return true;
                case ExpressionType.Convert:
                    var unaryExpression = (UnaryExpression)expression;
                    if (unaryExpression.Type.IsNullableType())
                        return GetValueFromExpressionWithoutConversion(unaryExpression.Operand, out value);
                    value = Expression.Lambda(expression).Compile().DynamicInvoke();
                    return true;
                case ExpressionType.NewArrayInit:
                    var expressions = ((NewArrayExpression)expression).Expressions;
                    var values = new object[expressions.Count];
                    value = null;
                    if (expressions.Where((t, i) => !GetValueFromExpressionWithoutConversion(t, out values[i])).Any())
                        return false;
                    value = values;
                    return true;
                case ExpressionType.NewArrayBounds:
                    value = null;
                    expressions = ((NewArrayExpression)expression).Expressions;
                    var constantExpression = (ConstantExpression)expressions.FirstOrDefault();
                    if (constantExpression == null)
                        return false;
                    if (constantExpression.Value.GetType() != typeof(int))
                        return false;
                    var length = (int)constantExpression.Value;
                    value = new object[length];
                    return true;
                default:
                    value = null;
                    return false;
            }
        }

        private static bool TryGetMethodArguments(MethodCallExpression mce, out object[] args)
        {
            args = new object[mce.Arguments.Count];
            for (var index = 0; index < mce.Arguments.Count; index++)
            {
                if (mce.Arguments[index].NodeType == ExpressionType.Lambda)
                {
                    args[index] = ((LambdaExpression)mce.Arguments[index]).Compile();
                    continue;
                }
                if (GetValueFromExpressionWithoutConversion(mce.Arguments[index], out var value) == false)
                    return false;
                args[index] = value;
            }
            return true;
        }

        private static object GetNewExpressionValue(Expression expression)
        {
            var newExpression = ((NewExpression)expression);
            var instance = Activator.CreateInstance(newExpression.Type, newExpression.Arguments.Select(e =>
            {
                object o;
                if (GetValueFromExpressionWithoutConversion(e, out o))
                    return o;
                throw new InvalidOperationException("Can't extract value from expression of type: " + expression.NodeType);
            }).ToArray());
            return instance;
        }

        private static object GetMemberValue(MemberExpression memberExpression)
        {
            object obj = null;

            if (memberExpression == null)
                throw new ArgumentNullException(nameof(memberExpression));

            // Get object
            if (memberExpression.Expression is ConstantExpression)
                obj = ((ConstantExpression)memberExpression.Expression).Value;
            else if (memberExpression.Expression is MemberExpression)
                obj = GetMemberValue((MemberExpression)memberExpression.Expression);
            else if (memberExpression.Expression is MethodCallExpression && GetValueFromExpressionWithoutConversion(memberExpression.Expression, out var value) &&
                     value is MethodCall mc)
            {
                mc.AccessPath = memberExpression.Member.Name;
                return mc;
            }
            //Fix for the issue here http://github.com/ravendb/ravendb/issues/#issue/145
            //Needed to allow things like ".Where(x => x.TimeOfDay > DateTime.MinValue)", where Expression is null
            //(applies to DateTime.Now as well, where "Now" is a property
            //can just leave obj as it is because it's not used below (because Member is a MemberInfo), so won't cause a problem
            else if (memberExpression.Expression != null)
                throw new NotSupportedException("Expression type not supported: " + memberExpression.Expression.GetType().FullName);

            if (obj is MethodCall m)
            {
                m.AccessPath += "." + memberExpression.Member.Name;
                return m;
            }
            // Get value
            var memberInfo = memberExpression.Member;
            if (memberInfo is PropertyInfo)
            {
                var property = (PropertyInfo)memberInfo;
                return property.GetValue(obj, null);
            }
            if (memberInfo is FieldInfo)
            {
                var value = Expression.Lambda(memberExpression).Compile().DynamicInvoke();
                return value;
            }
            throw new NotSupportedException("MemberInfo type not supported: " + memberInfo.GetType().FullName);
        }

        private static void AssertNoComputation(MemberExpression memberExpression)
        {
            var cur = memberExpression;

            while (cur != null)
            {
                switch (cur.Expression?.NodeType)
                {
                    case ExpressionType.Call:
                    case ExpressionType.Invoke:
                    case ExpressionType.Add:
                    case ExpressionType.And:
                    case ExpressionType.AndAlso:
                    case ExpressionType.AndAssign:
                    case ExpressionType.Decrement:
                    case ExpressionType.Increment:
                    case ExpressionType.PostDecrementAssign:
                    case ExpressionType.PostIncrementAssign:
                    case ExpressionType.PreDecrementAssign:
                    case ExpressionType.PreIncrementAssign:
                    case ExpressionType.AddAssign:
                    case ExpressionType.AddAssignChecked:
                    case ExpressionType.AddChecked:
                    case ExpressionType.Index:
                    case ExpressionType.Assign:
                    case ExpressionType.Block:
                    case ExpressionType.Conditional:
                    case ExpressionType.ArrayIndex:
                    case null:

                        throw new ArgumentException("Not supported computation: " + memberExpression +
                                                    ". You cannot use computation in RavenDB queries (only simple member expressions are allowed).");
                }
                cur = cur.Expression as MemberExpression;
            }
        }

        internal static string RemoveTransparentIdentifiersIfNeeded(string path)
        {
            while (path.StartsWith(JavascriptConversionExtensions.TransparentIdentifier))
            {
                var indexOf = path.IndexOf(".", StringComparison.Ordinal);
                path = path.Substring(indexOf + 1);
            }

            return path;
        }

        public static bool IsCounterCall(MethodCallExpression mce)
        {
            return mce.Method.DeclaringType == typeof(RavenQuery) && mce.Method.Name == "Counter"
                   || mce.Object?.Type == typeof(ISessionDocumentCounters) && mce.Method.Name == "Get";
        }

        public static bool IsTimeSeriesCall(MethodCallExpression mce)
        {
            var type = mce.Method.DeclaringType;

            if (type != null && type.IsGenericType)
                type = type.GetGenericTypeDefinition();
            
            if (type == typeof(ITimeSeriesQueryable) ||
                type == typeof(ITimeSeriesAggregationQueryable) ||
                type == typeof(ITimeSeriesQueryable<>) ||
                type == typeof(ITimeSeriesAggregationQueryable<>) ||
                mce.Method.ReturnType == typeof(ITimeSeriesQueryable) || // not a valid TimeSeries call expression
                mce.Method.ReturnType.IsGenericType &&
                mce.Method.ReturnType.GetGenericTypeDefinition() == typeof(ITimeSeriesQueryable<>)) // not a valid TimeSeries call expression
            {
                if (mce.Method.ReturnType == typeof(TimeSeriesRawResult) ||
                    mce.Method.ReturnType == typeof(TimeSeriesAggregationResult))
                    return true;

                if (mce.Method.ReturnType.IsGenericType)
                {
                    var definition = mce.Method.ReturnType.GetGenericTypeDefinition();
                    if (definition == typeof(TimeSeriesRawResult<>) || 
                        definition == typeof(TimeSeriesAggregationResult<>))
                        return true;
                }

                ThrowInvalidTimeSeriesReturnType(mce.Method.ReturnType);
            }

            return false;
        }

        private static void ThrowInvalidTimeSeriesReturnType(Type type)
        {
            throw new InvalidOperationException($"Time Series query expressions cannot return type '{type}'. " +
                                                $"Time Series query expressions must return type '{nameof(TimeSeriesRawResult)}' or '{nameof(TimeSeriesAggregationResult)}'. " +
                                                "Did you forget to call 'ToList'?");
        }
    }
}
